### PR TITLE
Fitst working version of storage builders/combinators.

### PR DESCRIPTION
**Pull request by alex-kooper**

### CODE REVIEW

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        return { ...selfStorage, ...otherStorage };
+      };
+      return storageBuilder(newBuild);
+    }
+  };
+
+  return self;
+};
+
+export const SimpleAdmin = storageBuilder(({ owner }: { owner: string }) => ({
```
Comment (emishur) [inline packages/fa2-interfaces/src/interfaces/storage-builder.ts:28]: owner should have `address` type

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  admin: {
+    admin: owner,
+    pending_admin: undefined,
+    paused: false
+  }
+}));
+
+export const Assets = storageBuilder(() => ({
```
Comment (emishur) [inline packages/fa2-interfaces/src/interfaces/storage-builder.ts:36]: There would be three different implementations of `assets`.
Consider naming them properly (nft, fungible, multifungible).

**Review by emishur — APPROVED** [review]

File: packages/fa2-interfaces/src/interfaces/admin-storage.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { MichelsonMap } from '@taquito/taquito';
+import { storageBuilder } from './storage-builder';
+import { address, unit } from '../type-aliases';
+
+const Simple = storageBuilder(({ owner }: { owner: string }) => ({
```
Comment (emishur) [inline packages/fa2-interfaces/src/interfaces/admin-storage.ts:5]: `owner: address`

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { MichelsonMap } from '@taquito/taquito';
+import { address, unit, nat } from '../type-aliases';
+import { TokenMetadataInternal } from './fa2';
+
+export interface StorageBuilder<I, S> {
+  build: (params: I) => S;
+  with: <I1, S1>(sb: StorageBuilder<I1, S1>) => StorageBuilder<I & I1, S & S1>;
+  transfer: <S1>(f: (s: S) => S1) => StorageBuilder<I, S1>;
```
Comment (emishur) [inline packages/fa2-interfaces/src/interfaces/storage-builder.ts:8]: What does `transfer` do? is it `map()` functor?
Consider adding doc commends and rename it to not confuse with FA2 `transfer`

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  return self;
+};
+
+export const Assets = storageBuilder(() => ({
+  assets: {
+    ledger: new MichelsonMap<nat, address>(),
+    operators: new MichelsonMap<[address, [address, nat]], unit>(),
+    token_metadata: new MichelsonMap<nat, TokenMetadataInternal>(),
+    mint_freeze: false
```
Comment (emishur) [inline packages/fa2-interfaces/src/interfaces/storage-builder.ts:38]: `mint_freeze` does not belong here as we discussed

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts (hunk 1 above)
Comment (alex-kooper) [reply packages/fa2-interfaces/src/interfaces/storage-builder.ts]: Done

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts (hunk 2 above)
Comment (alex-kooper) [reply packages/fa2-interfaces/src/interfaces/storage-builder.ts]: Done

File: packages/fa2-interfaces/src/interfaces/admin-storage.ts (hunk 3 above)
Comment (alex-kooper) [reply packages/fa2-interfaces/src/interfaces/admin-storage.ts]: done

File: packages/fa2-interfaces/src/interfaces/storage-builder.ts (hunk 5 above)
Comment (alex-kooper) [reply packages/fa2-interfaces/src/interfaces/storage-builder.ts]: Done

**Review by emishur — APPROVED** [review]

**Pull request merged**